### ISSUE TITLE
Possibly incorrect mean pooling

### ISSUE THREAD

**Issue by ir2718** [SeanLee97/AnglE#101]
Hi @SeanLee97 ,

I was going through the paper [AoE: Angle-optimized Embeddings for Semantic Textual Similarity](https://aclanthology.org/2024.acl-long.101.pdf), and I noticed in section 5.2 the following sentence: "Second, we test AoE’s performance on four different pooling strategies and find that the “cls” pooling is the most helpful one.". This struck me as very odd, since I've personally seen consistently better performance with mean pooling, as well in other related work. Naturally, I wanted to see what's going on so I took a look at the code, and compared it to sentence transformers code for pooling.

Correct me if I'm wrong, but your code does mean pooling like [this](https://github.com/SeanLee97/AnglE/blob/main/angle_emb/angle.py#L274):
```python
    elif pooling_strategy == 'avg':
        outputs = torch.sum(
            outputs * inputs["attention_mask"][:, :, None], dim=1) / torch.sum(inputs["attention_mask"])
```

However, this seems weird to me, as you're summing the whole attention mask into a single value. [This](https://huggingface.co/sentence-transformers/all-mpnet-base-v2#usage-huggingface-transformers) is the sentence transformers implementation:
```python
#Mean Pooling - Take attention mask into account for correct averaging
def mean_pooling(model_output, attention_mask):
    token_embeddings = model_output[0] #First element of model_output contains all token embeddings
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)
```

Here the sum is only over the first dimension. So I wrote a short script to compare the two methods, to make sure I'm not crazy:
```python
from transformers import AutoTokenizer, AutoModel
import torch

# SENT TF
def mean_pooling(model_output, attention_mask):
    token_embeddings = model_output[0] #First element of model_output contains all token embeddings
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)

# AoE
def aoe_pooling(outputs, inputs):
    outputs = torch.sum(outputs * inputs["attention_mask"][:, :, None], dim=1) / torch.sum(inputs["attention_mask"])
    return outputs

model_name = "bert-base-cased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)

tok = tokenizer(["a sentence", "just another sentence"], padding=True, return_tensors="pt")
output = model(**tok)

out1 = mean_pooling(output, tok["attention_mask"])
out2 = aoe_pooling(output.last_hidden_state, tok)

torch.set_printoptions(threshold=100)
print(out1)
print(out2)
```

The code does work for a single sentence in a batch, but for multiple sentences this gives different results:
```python
# only ["a sentence"]
tensor([[ 0.2745, -0.0422, -0.0991,  ..., -0.0648,  0.6051,  0.0599]],
       grad_fn=<DivBackward0>)
tensor([[ 0.2745, -0.0422, -0.0991,  ..., -0.0648,  0.6051,  0.0599]],
       grad_fn=<DivBackward0>)

# both ["a sentence", "just another sentence"]
tensor([[ 0.2745, -0.0422, -0.0991,  ..., -0.0648,  0.6051,  0.0599],
        [ 0.4062,  0.1066, -0.1681,  ...,  0.1155,  0.4484, -0.0193]],
       grad_fn=<DivBackward0>)
tensor([[ 0.1220, -0.0188, -0.0440,  ..., -0.0288,  0.2689,  0.0266],
        [ 0.2256,  0.0592, -0.0934,  ...,  0.0641,  0.2491, -0.0107]],
       grad_fn=<DivBackward0>)
```

Could you have a look and tell me if I missed something?

EDIT: 
In case it turns out to be incorrect, I opened up a PR: https://github.com/SeanLee97/AnglE/pull/102

**Comment by SeanLee97** [comment]
@ir2718 Thank you very much for your PR!

After reviewing the implementation, you are absolutely correct!
I forgot to set `keepdim=True` when summing the `attention_mask`, leading to a global sum being used.
